Add unit and E2E tests

AgnesToulet opened the pull request:
Fixes https://github.com/grafana/grafana-image-renderer/issues/434

evictorero commented on src/service/http-server.integration.test.ts:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  let diff: { width: number; height: number; data: Uint8ClampedArray } | null = null;
+  if (process.env['SAVE_DIFF'] === 'true') {
```
I saw we are setting some env files at the beginning inside `beforeAll` maybe we can read the env value there only once and create a variable to reference it here, similar to how we are setting up `domain`.

AgnesToulet replied:
I updated the code to do that 👍

evictorero commented on src/service/http-server.integration.test.ts:
```diff
@@ -0,0 +1,232 @@
+      maxConcurrency: 5,
+      timeout: 30,
+    },
+    verboseLogging: true,
```
The successful output is quite verbose. I tried changing this value to false, but it still logs a lot of debug messages. Is it possible to keep it less verbose by default?

AgnesToulet replied:
Yes I updated the config to remove some logs.

evictorero reviewed (commented):
I left some comments with thoughts and ideas but it works like a charm.

AgnesToulet commented on package.json:
```diff
@@ -17,7 +17,10 @@
-    "create-gcom-plugin-json": "ts-node scripts/createGcomPluginJson.ts ./scripts/tmp"
+    "create-gcom-plugin-json": "ts-node scripts/createGcomPluginJson.ts ./scripts/tmp",
+    "test": "jest",
+    "test-update": "cross-env UPDATE_GOLDEN=true jest",
```
The golden files are the expected output of the tests. You can find more info here for example: https://pkg.go.dev/github.com/xorcare/golden#section-readme

AgnesToulet commented on src/service/http-server.integration.test.ts:
```diff
@@ -0,0 +1,232 @@
+    expect(pixelDiff).toBeLessThan(imageDiffThreshold);
+  });
+
+  // TODO: this doesn't work on Drone
```
I removed the code and the expected image. I think we should test it but this is more complicated than the other and I don't think it should block this PR.

AgnesToulet commented on src/service/http-server.integration.test.ts:
```diff
@@ -0,0 +1,232 @@
+// It writes the diff file to /testdata if tests are run with SAVE_DIFF=true.
+function compareImage(testName: string, responseBody: any): number {
+  const goldenFilePath = path.join(goldenFilesFolder, `${testName}.png`);
+  if (process.env['UPDATE_GOLDEN'] === 'true') {
```
What about updating the `test-update` command to: 
```
"test-update": "cross-env UPDATE_GOLDEN=true jest -t=screenshot",
```
I'm worried that if we create a separate script, we will have a lot of duplicated code with the test. But I agree that we don't need to run all the tests so this would only run the screenshot test. We could do the same for the `test-diff` command.

evictorero replied:
The update script should be a loop over a map with a request and writing the response to an image. I think it should be decoupled entirely of the test suite. Maybe sounds like duplicating because that is what primarily tests are, just a request and comparing the response.
The map could contain a URL (the one used to render a panel) and an image name.
However, this shouldn't be a blocker, maybe this is too much and we can move on with this PR that is a great start and refactor later if needed.

AgnesToulet commented on package.json:
```diff
@@ -17,7 +17,10 @@
     "start": "node build/app.js server --config=dev.json",
-    "create-gcom-plugin-json": "ts-node scripts/createGcomPluginJson.ts ./scripts/tmp"
+    "create-gcom-plugin-json": "ts-node scripts/createGcomPluginJson.ts ./scripts/tmp",
+    "test": "jest",
```
I embedded running the Docker container in the `yarn test` command (I also created a new `yarn test-ci command for the CI), it should be easier to run tests locally now!

AgnesToulet commented:
@evictorero Thanks a lot for your thorough review! I updated the PR according to your feedback. I also had an issue with the Drone pipeline so I added a docker-compose file and some instructions to run tests in a Drone-like environment locally so this is easier to troubleshoot. 

It should work as before but please do test again to ensure I didn't break anything on non-Windows environments.

evictorero reviewed (approved):
Looks great! Left one comment to think about it.
I think it's ready for merge but only thing that bothers me a bit is the test output, when successful, it writes a lot of `console.log` messages and it confuses me because the tests are all successful.
Can we remove the output messages to avoid noise when all tests succeed?

AgnesToulet commented:
I have hidden the logs when tests are successful 👍 

For the update, I agree with you that it could be better, but I think it will make more sense to do it when we have more tests.
